### ISSUE TITLE
Bug: There are some cases that all View/Manager/Downloader completionBlock will not callback

### ISSUE THREAD

**Issue by dreampiggy** [SDWebImage/SDWebImage#2702]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | 12.0
 SDWebImage Version      | Both  5.0.1 and 4.4.6
 Integration Method      | cocoapods
 Xcode Version           | Xcode 10
 Repro rate              | sometimes 1%
 Repro with our demo prj | does it happen with our demo project? YES
 Demo project link       | demo project

### Issue Description and Steps

SDWebImage have some completion block to notify the async loading task been finished, including

+ `-[UIImageView sd_setImageWithURL:completed]`: `SDExternalCompletionBlock`
+ `-[SDWebImageManager loadImageWithURL:]`: `SDInternalCompletionBlock`
+ `-[downloadImageWithURL downloadImageWithURL:]`: `SDWebImageDownloaderCompletedBlock`

But sometimes, the all of the above completion block, is not get callback, which make the users's App become unexpceted. For example, the code we use like this:

```objectivec
UIImageView *imageView;
imageView.sd_indicator = SDWebImageActivityIndicator.grayIndicator;
[imageView sd_setImageWithURL:url completed:^(UIImage * _Nullable image, NSError * _Nullable error, SDImageCacheType cacheType, NSURL * _Nullable imageURL) {
    // never callback, and the indicator keep spining
}];

// later some other code trigger this
[SDWebImageDownloader cancelAllDownloads];
```

If this case happended, some logic related to the completion is now been skipped and cause unexpceted behavior for user's App.

This is just one case I found that can cause the completion block not been callback, there may be some other cases, but at least this one is easy to reproduce.

I think, in each cases, the image loading task should report to the user, if it's cancelled, it still should be callbacked with specify error. Because it's a asynchronous operation with some return result (the image, data, cache type and error args), but not synchronous with the return value.

### Kingfisher

I also try another popular image loading framework [Kingfisher](https://github.com/onevcat/Kingfisher). Which does the correct thing to callback even it's canncelled. 
I think they do the correct behavior for this case.

### Related

#1209 #2062

**Comment by zhongwuzw** [comment]
> https://github.com/SDWebImage/SDWebImage/pull/2692#issuecomment-486069081

@dreampiggy I list one of them:
1. Let's imagine a concurrency load images.
2. User loads image `url1`, and downloading by network.
3. `url1` downloading continuing, now user loads image `url1` again, and code going to but not yet to execute this line
https://github.com/SDWebImage/SDWebImage/blob/6c1688962869f51e787e25f123b7f9b482368b00/SDWebImage/SDWebImageDownloader.m#L244, before it, the downloading operation finished execute this line 
https://github.com/SDWebImage/SDWebImage/blob/master/SDWebImage/SDWebImageDownloaderOperation.m#L499, already get the callbacks block array, so now even `addHandlersForProgress:completed` be called, it's too late to be added to callback array.

**Comment by Speakus** [comment]
I care about issue

**Comment by dreampiggy** [comment]
Fixed by #2692.
